### ISSUE TITLE
why both train_perplexity and train_epoch_loss are nan 

### ISSUE THREAD

**Issue by judyhappy** [facebookresearch/llama-recipes#300]
### System Info

Tesla V100

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

I tried to fine tune the llama2 7B model with my dataset.
Both train_perplexity and  train_epoch_loss are always nan during fine tune progress.
what is the possible reason?

### Error logs

Training Epoch: 4/150, step 63/64 completed (loss: 0.19040994346141815): 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 64/64 [04:02<00:00,  3.78s/it]
Max CUDA memory allocated was 1 GB
Max CUDA memory reserved was 1 GB
Peak active CUDA memory was 1 GB
Cuda Malloc retires : 0
CPU Total Peak Memory consumed during the train (max): 14 GB
evaluating Epoch: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [01:53<00:00,  1.77it/s]
 eval_ppl=tensor(nan, device='cuda:0') eval_epoch_loss=tensor(nan, device='cuda:0')
Epoch 4: train_perplexity=nan, train_epoch_loss=nan, epoch time 242.3404811769724s
Training Epoch: 5/150, step 63/64 completed (loss: 0.14372023940086365): 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 64/64 [04:01<00:00,  3.78s/it]
Max CUDA memory allocated was 1 GB
Max CUDA memory reserved was 1 GB
Peak active CUDA memory was 1 GB
Cuda Malloc retires : 0
CPU Total Peak Memory consumed during the train (max): 14 GB

### Expected behavior

train_perplexity and train_epoch_loss should't be NAN

**Comment by HamidShojanazeri** [comment]
@judyhappy can you pls share what command you are running? and more on your setting.

**Comment by wukaixingxp** [comment]
Hi! It seems that there has not been a follow-up conversation for a long time. I will close this issue for now and feel free to reopen it if you have any questions!

**Comment by smartinezai** [comment]
I have the same issue. I am trying to run finetuning.py on a custom dataset using llama 3.2 1b instruct and here's what bI get 

```
torchrun /leonardo_work/EUHPC_A04_062/AI_Model/src/llama_cookbook/finetuning.py \
>     --enable_fsdp \
>     --model_name /leonardo_work/EUHPC_A04_062/Llama-3.2-1B-Instruct \
>     --dataset monke_dataset \
>     --save_model \
>     --dist_checkpoint_root_folder model_checkpoints \
>     --dist_checkpoint_folder fine-tuned \
>     --fsdp_config.pure_bf16 \
> print("3"
-bash: syntax error near unexpected token `('
(venv_project) [smartin2@lrdn2081 ~]$ torchrun /leonardo_work/EUHPC_A04_062/AI_Model/src/llama_cookbook/finetuning.py     --enable_fsdp     --model_name /leonardo_work/EUHPC_A04_062/Llama-3.2-1B-Instruct     --dataset monke_dataset     --save_model     --dist_checkpoint_root_folder model_checkpoints     --dist_checkpoint_folder fine-tuned     --fsdp_config.pure_bf16






/leonardo_work/EUHPC_A04_062/AI_Model/src/llama_cookbook/model_checkpointing/checkpoint_handler.py:17: DeprecationWarning: `torch.distributed._shard.checkpoint` will be deprecated, use `torch.distributed.checkpoint` instead
  from torch.distributed._shard.checkpoint import (
Clearing GPU cache for all ranks
--> Running with torch dist debug set to detail
--> Model /leonardo_work/EUHPC_A04_062/Llama-3.2-1B-Instruct

--> /leonardo_work/EUHPC_A04_062/Llama-3.2-1B-Instruct has 1235.8144 Million params

bFloat16 enabled for mixed precision - using bfSixteen policy
/leonardo_work/EUHPC_A04_062/venv_project/lib/python3.11/site-packages/torch/distributed/fsdp/_init_utils.py:444: UserWarning: FSDP is switching to use `NO_SHARD` instead of ShardingStrategy.FULL_SHARD since the world size is 1.
  warnings.warn(
--> applying fsdp activation checkpointing...
--> Training Set Length = 100
--> Validation Set Length = 100
Preprocessing dataset: 100%|███████████████████████████████████████████████████| 100/100 [00:00<00:00, 535.44it/s]
length of dataset_train 2
--> Num of Training Set Batches loaded = 0
Preprocessing dataset: 100%|██████████████████████████████████████████████████| 100/100 [00:00<00:00, 3693.05it/s]
--> Num of Validation Set Batches loaded = 2
--> Num of Validation Set Batches loaded = 2
Starting epoch 0/3
train_config.max_train_step: 0
/leonardo_work/EUHPC_A04_062/venv_project/lib/python3.11/site-packages/torch/cuda/memory.py:391: FutureWarning: torch.cuda.reset_max_memory_allocated now calls torch.cuda.reset_peak_memory_stats, which resets /all/ peak memory stats.
  warnings.warn(
Training Epoch: 1: 0it [00:00, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Training Epoch: 1: 0it [00:00, ?it/s]
Max CUDA memory allocated was 2 GB
Max CUDA memory reserved was 2 GB
Peak active CUDA memory was 2 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 4 GB
evaluating Epoch:   0%|                                                                     | 0/2 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
evaluating Epoch: 100%|█████████████████████████████████████████████████████████████| 2/2 [00:13<00:00,  6.93s/it]
 eval_ppl=tensor(3.0533, device='cuda:0') eval_epoch_loss=tensor(1.1162, device='cuda:0')
 Saving the FSDP model checkpoints and optimizer using SHARDED_STATE_DICT
=====================================================
Saving model to /leonardo/home/userexternal/smartin2/model_checkpoints/fine-tuned-/leonardo_work/EUHPC_A04_062/Llama-3.2-1B-Instruct
/leonardo_work/EUHPC_A04_062/venv_project/lib/python3.11/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py:690: FutureWarning: FSDP.state_dict_type() and FSDP.set_state_dict_type() are being deprecated. Please use APIs, get_state_dict() and set_state_dict(), which can support different parallelisms, FSDP1, FSDP2, DDP. API doc: https://pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict.get_state_dict .Tutorial: https://pytorch.org/tutorials/recipes/distributed_checkpoint_recipe.html .
  warnings.warn(
/leonardo_work/EUHPC_A04_062/venv_project/lib/python3.11/site-packages/torch/distributed/fsdp/_state_dict_utils.py:773: UserWarning: When using ``NO_SHARD`` for ``ShardingStrategy``, full_state_dict willbe returned.
  warnings.warn(
/leonardo_work/EUHPC_A04_062/venv_project/lib/python3.11/site-packages/torch/distributed/fsdp/_state_dict_utils.py:711: UserWarning: When using ``NO_SHARD`` for ``ShardingStrategy``, full_state_dict willbe returned.
  warnings.warn(
/leonardo_work/EUHPC_A04_062/AI_Model/src/llama_cookbook/model_checkpointing/checkpoint_handler.py:113: FutureWarning: `save_state_dict` is deprecated and will be removed in future versions.Please use `save` instead.
  dist_cp.save_state_dict(
Sharded state checkpoint saved to /leonardo/home/userexternal/smartin2/model_checkpoints/fine-tuned-/leonardo_work/EUHPC_A04_062/Llama-3.2-1B-Instruct
Checkpoint Time = 4.0732

best eval loss on epoch 1 is 1.1162092685699463
Epoch 1: train_perplexity=nan, train_epoch_loss=nan, epoch time 0.801575162447989s
Starting epoch 1/3
train_config.max_train_step: 0
Training Epoch: 2: 0it [00:00, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Training Epoch: 2: 0it [00:00, ?it/s]
Max CUDA memory allocated was 2 GB
Max CUDA memory reserved was 2 GB
Peak active CUDA memory was 2 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 12 GB
evaluating Epoch:   0%|                                                                     | 0/2 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
evaluating Epoch: 100%|█████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  7.27it/s]
 eval_ppl=tensor(3.0533, device='cuda:0') eval_epoch_loss=tensor(1.1162, device='cuda:0')
Epoch 2: train_perplexity=nan, train_epoch_loss=nan, epoch time 0.3726232061162591s
Starting epoch 2/3
train_config.max_train_step: 0
Training Epoch: 3: 0it [00:00, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Training Epoch: 3: 0it [00:00, ?it/s]
Max CUDA memory allocated was 2 GB
Max CUDA memory reserved was 2 GB
Peak active CUDA memory was 2 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 12 GB
evaluating Epoch:   0%|                                                                     | 0/2 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
evaluating Epoch: 100%|█████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  7.59it/s]
 eval_ppl=tensor(3.0533, device='cuda:0') eval_epoch_loss=tensor(1.1162, device='cuda:0')
Epoch 3: train_perplexity=nan, train_epoch_loss=nan, epoch time 0.3641195297241211s
training params are saved in /leonardo/home/userexternal/smartin2/model_checkpoints/fine-tuned-/leonardo_work/EUHPC_A04_062/Llama-3.2-1B-Instruct/train_params.yaml
Key: avg_train_prep, Value: nan
Key: avg_train_loss, Value: nan
Key: avg_eval_prep, Value: 3.053258180618286
Key: avg_eval_loss, Value: 1.1162092685699463
Key: avg_epoch_time, Value: 0.5127726327627897
Key: avg_checkpoint_time, Value: 1.3812162109340231
```
